packages/expo-router/src/global-state/routing.ts

```diff
@@ -167,3 +178,28 @@ function getNavigateAction(state: ResultState, parentState: NavigationState, typ
     },
   };
 }
+
+function areRoutesEqual(
+  a: PartialRoute<any> | Route<any> = {},
+  b: PartialRoute<any> | Route<any> = {}
+) {
+  // If params if the names are different
+  if (a.name !== b.name) return false;
+
+  const paramsA = a.params;
+  const paramsB = b.params;
+
+  // If both null return true
+  if (paramsA === paramsB) return true;
+  // If one is null, return false
+  if (!paramsA || !paramsB) return false;
+
+  // Otherwise compare both params objects
+  const keysA = Object.keys(paramsA);
+
+  return keysA.every((key) => {
+    const valueA = paramsA[key];
+    const valueB = paramsB[key];
+    return valueA === valueB || valueA?.toString?.() === valueB?.toString?.();
```

Comment:
```suggestion
    return valueA == valueB
```

If we're comparing numbers and strings, it would be cheaper to use `==` than to optionally convert to string.